samples/modules/thrift/hello/client/src/main.cpp

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2022 Young Mei
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifdef __ZEPHYR__
+#include <zephyr/kernel.h>
+#endif
+
+#include <cstdlib>
+#include <iostream>
+
+#include <fcntl.h>
+
+#include <thrift/protocol/TBinaryProtocol.h>
+#include <thrift/protocol/TCompactProtocol.h>
+#include <thrift/transport/TBufferTransports.h>
+#include <thrift/transport/TSSLSocket.h>
+#include <thrift/transport/TSocket.h>
+
+#include "Hello.h"
+
+using namespace apache::thrift;
+using namespace apache::thrift::protocol;
+using namespace apache::thrift::transport;
+
+#ifndef IS_ENABLED
+#define IS_ENABLED(flag) flag
+#endif
+
+#ifndef CONFIG_THRIFT_COMPACT_PROTOCOL
+#define CONFIG_THRIFT_COMPACT_PROTOCOL 0
+#endif
+
+#ifndef CONFIG_THRIFT_SSL_SOCKET
+#define CONFIG_THRIFT_SSL_SOCKET 0
+#endif
+
+#ifdef __ZEPHYR__
+int main(void)
+#else
+int main(int argc, char **argv)
+#endif
+{
+	std::string my_addr;
+
+#ifdef __ZEPHYR__
+	my_addr = CONFIG_NET_CONFIG_PEER_IPV4_ADDR;
+#else
+	if (IS_ENABLED(CONFIG_THRIFT_SSL_SOCKET)) {
+		if (argc != 5) {
+			printf("usage: %s <ip> <native-cert.pem> <native-key.pem> "
+			       "<qemu-cert.pem>\n",
+			       argv[0]);
+			return EXIT_FAILURE;
+		}
+	}
+
+	if (argc >= 2) {
+		my_addr = std::string(argv[1]);
+	} else {
+		my_addr = "192.0.2.1";
+	}
+#endif
+
+	int port = 4242;
+	std::shared_ptr<TProtocol> protocol;
+	std::shared_ptr<TTransport> transport;
+	std::shared_ptr<TSSLSocketFactory> socketFactory;
+	std::shared_ptr<TTransport> trans;
+
+	if (IS_ENABLED(CONFIG_THRIFT_SSL_SOCKET)) {
+		const int port = 4242;
+		socketFactory = std::make_shared<TSSLSocketFactory>();
+		socketFactory->authenticate(true);
+
+#ifdef __ZEPHYR__
+		static const char qemu_cert_pem[] = {
+#include "qemu_cert.pem.inc"
+		};
+
+		static const char qemu_key_pem[] = {
+#include "qemu_key.pem.inc"
+		};
+
+		static const char native_cert_pem[] = {
+#include "native_cert.pem.inc"
+		};
```

Comment:
I don't like this because these will only get updated if this .cpp file is updated, if someone changes the .pem files, it will look like the application has been updated but it is a silent failure because nothing has changed in this file and there is no linkage between the .pem files and this .cpp file - so the user will be left wondering why their custom certificates do not seem to be working

Reply:
Please re-evaluate your comment, and again, be a bit more specific about requesting an actual change if there is a specific change that should be made.

You already requested that `.pem` files not be included here directly, as you hypothesized that changing a `.pem` file might not trigger a rebuild if changed.

In any case, last week I made significant change to the structure of the samples and used the common Zephyr way of generating a `.inc` file out of a `.pem` file. 

There are cmake build rules now that explicitly track dependencies (although, I'm suspect the dependency would be tracked before as well anyway).

Reply:
Indeed it seems cmake is smart enough to deal with this (the reason why it caused me concern was with just a makefile based project whereby the dependencies are not taken into consideration, and a .c file that was compiled included other .c files that were not compiled on their own, if the other .c files were changed and an incremental build was made, nothing would actually be rebuilt)

Reply:
I see - it should not be a concern for the Makefile project, as that is outside the scope of the `#ifdef __ZEPHYR__` block.